### PR TITLE
DM-22796: Rearrange test so that temp data files are removed if tests pass

### PR DESCRIPTION
**Pull request by timj**
This saves 200MB in the installed footprint.  The test data are
not deleted if a test fails.

### CODE REVIEW

File: tests/nopytest_test_coadds.py [hunk 1]
```diff
@@ -80,72 +80,24 @@
 DATAREPO_ROOT = os.path.join(os.path.dirname(__file__), ".tests", "testCoadds-data")
 
 
+def assertWrapper(func):
+    """Decorator to intercept AssertionError and reraise whilst recording
+    a failure."""
+    def wrapped(self):
+        try:
+            func(self)
+        except Exception:
```
Comment (jdswinbank) [inline tests/nopytest_test_coadds.py]: Should we be catching `Exception` here, or `AssertionError` (per the docstring, above)?

File: tests/nopytest_test_coadds.py [hunk 2]
```diff
@@ -218,6 +170,85 @@ def runForcedPhotCcdTask(butler):
...
+        runForcedPhotCoaddTask(butler, mocksTask)
+        runForcedPhotCcdTask(butler)
+
+    @classmethod
+    def tearDownClass(cls):
+        """Removes test data if tests ran okay"""
+        if os.path.exists(DATAREPO_ROOT):
+            if not cls.failed:
+                print(f"Deleting temporary data repository {DATAREPO_ROOT}")
+                # Errors can be ignored at this point
```
Comment (jdswinbank) [inline tests/nopytest_test_coadds.py]: I'm not sure the comment really adds much... presumably, if the errors couldn't be ignored, the code wouldn't say `ignore_errors=True`. :-) I'd suggest either removing the comment, or using it to explain _why_ it's appropriate to ignore errors.

File: tests/nopytest_test_coadds.py [hunk 3]
```diff
@@ -218,6 +170,85 @@ def runForcedPhotCcdTask(butler):
...
+        mergeMeasTask = MergeMeasurementsTask(config=mergeMeasConfig, butler=butler)
+        mergeMeasTask.writeSchemas(butler)
+        runTaskOnPatchList(butler, mergeMeasTask, mocksTask)
+
+        runForcedPhotCoaddTask(butler, mocksTask)
+        runForcedPhotCcdTask(butler)
+
+    @classmethod
+    def tearDownClass(cls):
+        """Removes test data if tests ran okay"""
```
Comment (jdswinbank) [inline tests/nopytest_test_coadds.py]: Perhaps “if all tests passed” would be clearer?

File: tests/nopytest_test_coadds.py [hunk 4]
```diff
@@ -80,72 +80,24 @@
 DATAREPO_ROOT = os.path.join(os.path.dirname(__file__), ".tests", "testCoadds-data")
 
 
+def assertWrapper(func):
+    """Decorator to intercept AssertionError and reraise whilst recording
+    a failure."""
```
Comment (jdswinbank) [inline tests/nopytest_test_coadds.py]: I think it'd be worth writing just another sentence here describing why this is a useful thing to do, just to help the reader orient themselves. Something along the lines of “this enables us to track whether any tests failed, and if so to preserve the test data for future reference”.

File: tests/nopytest_test_coadds.py (hunk 1 above)
Comment (timj) [reply tests/nopytest_test_coadds.py]: I did have AssertionError here in the lsst_ci variant but then I got to pondering that any failures whilst running a test would also be a good reason not to delete. I can fix the docstring.

**Pull request merged**